File: chapters/05_retry_and_recovery/v1_fail_stop.ts
// TypeScript mirror for Chapter 05: Retry and Recovery (v1)
// Run with: npx ts-node v1_fail_stop.ts

function runOnce(simulatedError?: string): [string, string[]] {
  const logs: string[] = ["start"];
  if (simulatedError) {
    logs.push(`error:${simulatedError}`);
    logs.push("stop");
    return ["failed", logs];
  }

  logs.push("success");
  return ["success", logs];
}

function main(): void {
  console.log(runOnce("timeout"));
}

main();

File: chapters/05_retry_and_recovery/v1_fail_stop.py
from __future__ import annotations


def run_once(simulated_error: str | None = None) -> tuple[str, list[str]]:
    logs: list[str] = ["start"]
    if simulated_error:
        logs.append(f"error:{simulated_error}")
        logs.append("stop")
        return "failed", logs

    logs.append("success")
    return "success", logs


def main() -> None:
    print(run_once("timeout"))


if __name__ == "__main__":
    main()
